### ISSUE TITLE
Config file min volume

### ISSUE THREAD

**Issue by qkqsb** [RHVoice/RHVoice#594]
Hello,

I'm using RHVoice in conjunction with OsmAnd. When I drive I play music, and I'd like to be able to hear my driving instructions clearly from RHVoice while doing so. I would expect the instructions at a similar volume to that of the music, but RHVoice is substantially quieter. I know this behaviour is due to RHVoice and not OsmAnd because if you replace RHVoice with Google Speech Services, you can change a setting in speech services and get a reasonable volume out. I looked into the config file, and I was puzzled as the path indicated in [the documentation ](https://github.com/RHVoice/RHVoice/blob/master/doc/en/Configuration-file.md)does not exist. I created a config file, and the app prompted me for a path to the file, which I gave it. I set min_volume to both .25 and 1. I didn't hear much difference, so I'm not sure it worked. If it did in fact work, I'm wondering if there is some way to increase the volume past 1, as it is nowhere near loud enough.

**Comment by zstanecic** [comment]
Hi,

Have you removed the ; sign, semicolon before each of the respective parameters?

 

From: qkqsb ***@***.***> 
Sent: Wednesday, July 6, 2022 3:34 AM
To: RHVoice/RHVoice ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [RHVoice/RHVoice] Config file min volume (Issue #594)

 

Hello,

I'm using RHVoice in conjunction with OsmAnd. When I drive I play music, and I'd like to be able to hear my driving instructions clearly from RHVoice while doing so. I would expect the instructions at a similar volume to that of the music, but RHVoice is substantially quieter. I know this behaviour is due to RHVoice and not OsmAnd because if you replace RHVoice with Google Speech Services, you can change a setting in speech services and get a reasonable volume out. I looked into the config file, and I was puzzled as the path indicated in the documentation  <https://github.com/RHVoice/RHVoice/blob/master/doc/en/Configuration-file.md> does not exist. I created a config file, and the app prompted me for a path to the file, which I gave it. I set min_volume to both .25 and 1. I didn't hear much difference, so I'm not sure it worked. If it did in fact work, I'm wondering if there is some way to increase the volume past 1, as it is nowhere near loud enough.

**Comment by qkqsb** [comment]
Zvonimir,

I am fairly sure the config file is not being respected at all. There are
no semicolons. Here are the exact contents of my config.

min_volume=1
quality=max
default_rate=5
default_pitch=2

I can change the values for these settings all I like. Playing example
sounds via the RHVoice app or Android settings produces no change in output.

**Comment by qkqsb** [comment]
@zstanecic Following up on this, do you have any recommendations for me to troubleshoot?

**Comment by zstanecic** [comment]
Try to set the higher range of the volume: 0.25 minimal, 4.0 maximum.

**Comment by jamesericdavidson** [comment]
> Try to set the higher range of the volume: 0.25 minimal, 4.0 maximum.

Tried this, with no success.

Like @qkqsb, I can't access the `Android/data` directory from the RHVoice app.

I wrote an `RHVoice.conf` file to the `Android` directory, but it has no effect while using Organic Maps. The volume is too low to be heard over music.

Using GrapheneOS on Pixel 4a, for reference.

**Comment by qkqsb** [comment]
> Try to set the higher range of the volume: 0.25 minimal, 4.0 maximum.
@zstanecic This was not successful for me. Here is my config:

min_volume=1
max_volume=4
quality=max
min_rate=0.5
max_rate=5
default_rate=5
default_pitch=2
default_volume=4

I have a theory that if I use a path which the application is expecting instead of a user defined one, maybe it will be enforced seeing as the documentation previously mentioned references a specific path. I located what I believe to be the correct directory in /storage/self/primary/Android/data/com.github.olga_yakovleva.rhvoice.android/files and created a config file here. I decided to select this config as well within the app for good measure, and was surprised that the file was actually overwrote to be blank upon doing this. I copied in the contents of the config again, and tried experimenting with various settings to see if I could produce a noticeable change. Nothing happened when any setting was changed. I tried this on a second device as well, same result. 

I am unsure how else to proceed, but I do believe that the [documentation](https://github.com/RHVoice/RHVoice/blob/master/doc/en/Configuration-file.md) should be clarified to give some insight as to where this relative file path for the config starts.

**Comment by zstanecic** [comment]
Hi,,

You should import the config via settings.

**Comment by qkqsb** [comment]
@zstanecic 
As previously stated, I have been very deliberate about applying the config within the app. When the config is selected it will erase any contents of the file, rendering it useless. If one goes in after the config has been selected, and the app has retained the path to this file as indicated by the switch for the config in the settings actually being enabled, and then applies the contents of the config file again, nothing will change.

TLDR: Not only is the config file not being respected by the application even when placed in the location mentioned by the documentation, it is being erased once selected.

**Comment by qkqsb** [comment]
It looks like the features for a config file in Android were implemented in [this commit](https://github.com/RHVoice/RHVoice/commit/5b64df483788b9941b9a4053f3ed471635ee38f3) by @Olga-Yakovleva.
Apologies if what I say below is in any way incorrect or misinformed, my Java experience is almost as minuscule as my familiarity with this code base. I am simply trying to move this issue forward.
 It looks like the changes that were made to src/android/RHVoice-core/src/main/java/com/github/olga_yakovleva/rhvoice/android/SettingsFragment.java in this commit  are hardly present in the current version of the file. I particularly looked for the lines with volumePref, and see very little. I don't know if this is simply how the code has evolved or if I'm missing something.

If I am taking this in the wrong direction and simply need to do more testing, I would appreciate it if someone could offer some new and constructive ideas that have not yet been tried.

**Comment by qkqsb** [comment]
Note to those impacted by this: eSpeak NG can increase its output volume to 200%. While this is not perfect, it is a substantial improvement upon what RHVoice is capable of at the moment.

**Comment by qkqsb** [comment]
Changing the title to reflect current findings.

I built the application from source and used a debugging version of the apk. The configuration file is being enforced, and I can produce a perceptible change in pitch. However, I cannot produce a change in volume using the config file.

I modified /src/core/params.cpp line 20 to change MAX_MAX_VOLUME=4 to MAX_MAX_VOLUME=40 and rebuilt the app. No perceptible change still even when making changes to the config file again. I did however discover a volume slider in the app which doesn't seem to change the actual config file. I've included a screenshot of the slider found under Settings  > Languages > English > Speech volume here. There is no change to the copy of the config file within the app's data directory after changes are made to the slider. Changes to the slider do make noticeable changes in volume.

So, why two independent ways to change the volume? Why does the slider work, and the config file not? Expected behavior would be to have the slider change the config file.

![rhvoice_volume_slider](https://github.com/RHVoice/RHVoice/assets/90565973/7278c640-d989-4fa8-952c-ec651a73d95c)

**Comment by beqabeqa473** [comment]
hello @qkqsb. I think you are miss-understanding some points:
1. slider in application settings is not linked directly to RHVoice config.
2. inRHVoice config file there are two parameters, min_volume and max_volume, which will be rescaled in percents when you adjust volume from application settings.

If you still have problems with that, please give us detailed explanation, which phone model, android version is used.
If all is fixed for you, and this was just matter of miss-understanding, you can close this issue.
